Many experiments and data sets involve multiple values of the same type. An *array* is a collection of values that all have the same type. The `numpy` package, abbreviated `np` in programs, provides Python programmers with convenient and powerful functions for creating and manipulating arrays.

In [7]:
import numpy as np

An array is created using the `np.array` function, which takes a list or tuple as an argument. Here, we create arrays of average daily [high](http://berkeleyearth.lbl.gov/auto/Regional/TMAX/Text/global-land-TMAX-Trend.txt) and [low](http://berkeleyearth.lbl.gov/auto/Regional/TMIN/Text/global-land-TMIN-Trend.txt) temperatures for the decades surrounding 1850, 1900, 1950, and 2000.

In [15]:
baseline_high = 14.48
highs = np.array([baseline_high - 0.880, 
                  baseline_high - 0.093,
                  baseline_high + 0.105, 
                  baseline_high + 0.684])
highs

array([ 13.6  ,  14.387,  14.585,  15.164])

In [16]:
baseline_low = 3.00
lows = np.array((baseline_low - 0.872, baseline_low - 0.629,
                 baseline_low - 0.126, baseline_low + 0.728))
lows

array([ 2.128,  2.371,  2.874,  3.728])

Arrays differ from lists and tuples because they can be used in arithmetic expressions to compute over their contents. When an array is combined with a single number, that number is combined with each element of the array. Therefore, we can convert all of these temperatures to Farenheit by writing the familiar conversion formula.

In [10]:
(9/5) * highs + 32

array([ 56.48  ,  57.8966,  58.253 ,  59.2952])

When two arrays are combined together using an arithmetic operator, their individual values are combined. For example, we can compute the daily temperature ranges by subtracting.

In [11]:
highs - lows

array([ 11.472,  12.016,  11.711,  11.436])

Arrays also have *methods*, which are functions that operate on the array values. The `mean` of a collection of numbers is its average value: the sum divided by the length. Each pair of parentheses in the example below is part of a call expression; it's calling a function with no arguments to perform a computation on the array called `temps`.

In [17]:
[highs.size, highs.sum(), highs.mean()]

[4, 57.736000000000004, 14.434000000000001]

It's common that small rounding errors appear in arithmetic on a computer. For instance, the mean (average) above is in fact `14.434`, but an extra 0.0000000000000001 has been added during the computation of the average. Arithmetic rounding errors are an important topic in scientific computing, but won't be a focus of this text. Everyone who works with data must be aware that some operations will be approximated so that values can be stored and manipulated efficiently.

#### Functions on Arrays

In addition to basic arithmetic and methods such as `min` and `max`, manipulating arrays often involves calling functions that are part of the `np` package. For example, the `diff` function computes the difference between each adjacent pair of elements in an array. The first element of the `diff` is the second element minus the first. 

In [19]:
np.diff(highs)

array([ 0.787,  0.198,  0.579])

The [full Numpy reference](http://docs.scipy.org/doc/numpy/reference/) lists these functions exhaustively, but only a small subset are used commonly for data processing applications. These are grouped into different packages within `np`. Learning this vocabulary is an important part of learning the Python language, so refer back to this list often as you work through examples and problems.

Each of these functions takes an array as an argument and returns a single value.

| **Function**       | Description                                                          |
|--------------------|----------------------------------------------------------------------|
| `np.prod`          | Multiply all elements together                                       |
| `np.sum`           | Add all elements together                                            |
| `np.all`           | Test whether all elements are true values (non-zero numbers are true)|
| `np.any`           | Test whether any elements are true values (non-zero numbers are true)|
| `np.count_nonzero` | Count the number of non-zero elements                                |

Each of these functions takes an array as an argument and returns an array of values.

| **Function**       | Description                                                          |
|--------------------|----------------------------------------------------------------------|
| `np.diff`          | Difference between adjacent elements                                 |
| `np.round`         | Round each number to the nearest integer (whole number)              |
| `np.cumprod`       | A cumulative product: for each element, multiply all elements so far |
| `np.cumsum`        | A cumulative sum: for each element, add all elements so far          |
| `np.exp`           | Exponentiate each element                                            |
| `np.log`           | Take the natural logarithm of each element                           |
| `np.sqrt`          | Take the square root of each element                                 |
| `np.sort`          | Sort the elements                                                    |

Each of these functions takes an array of strings and returns an array.

| **Function**        | **Description**                                              |
|---------------------|--------------------------------------------------------------|
| `np.char.lower`     | Lowercase each element                                       |
| `np.char.upper`     | Uppercase each element                                       |
| `np.char.strip`     | Remove spaces at the beginning or end of each element        |
| `np.char.isalpha`   | Whether each element is only letters (no numbers or symbols) |
| `np.char.isnumeric` | Whether each element is only numeric (no letters)  

Each of these functions takes both an array of strings and a *search string*; each returns an array.

| **Function**         | **Description**                                                                  |
|----------------------|----------------------------------------------------------------------------------|
| `np.char.count`      | Count the number of times a search string appears among the elements of an array |
| `np.char.find`       | The position within each element that a search string is found first             |
| `np.char.rfind`      | The position within each element that a search string is found last              |
| `np.char.startswith` | Whether each element starts with the search string  

